### PR TITLE
Use multiple source connectors with the same table

### LINKED ISSUE #63: Feature: Update the naming of conduit items in the database
### Feature description

Add a suffix to the conduit items (e.g.: tracking table, trigger, etc.), so that allows multiple pipelines to be created for a single table.

### PR DESCRIPTION
**Pull request by hariso**
### Description

Fixes #63 and #70.

The approach is the following: to use multiple source connectors with the same table and at the same time, we need to have different snapshot table, tracking table, and trigger names. The names are generated once when the connector starts and then saved (in the position).

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/conduitio-labs/conduit-connector-oracle/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I ran the acceptance test and the integration tests on df23dff389d63fe825e0e7893ad9e8b7918ae4d2 (outdated)
- [x]  I ran the acceptance test and the integration tests on a6145b91233004f0c2d81a3e09cc4e08604ee3d3 (outdated)
- [ ] I ran the acceptance test and the integration tests on 60db0552bfbedb53bc9b3f2dc0068cf5720d3197
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: source/iterator/iterator.go [hunk 1]
```diff
@@ -114,20 +114,41 @@ type Params struct {
 	BatchSize      int
 }
 
+// HelperObjects returns the helper tables and triggers we use:
+// the snapshot table, the tracking table and the trigger name.
+func (p Params) HelperObjects() (string, string, string) {
```
Comment (hariso) [inline source/iterator/iterator.go:119]: I'm not really happy about this name, I'll try to come up with something better.

File: columntypes/columntypes.go [hunk 2]
```diff
@@ -116,6 +116,8 @@ func GetColumnTypes(
 	var columnName string
 
 	rows, err := repo.DB.QueryxContext(ctx, fmt.Sprintf(queryColumnDescription, tableName))
+	defer rows.Close() //nolint:staticcheck // error doesn't affect outcome
```
Comment (lovromazgon) [inline columntypes/columntypes.go:119]: We shouldn't close the rows before we check the error, rows might be `nil` (see [example](https://pkg.go.dev/database/sql#example-Rows)).

File: source/iterator/iterator.go (hunk 1 above)
Comment (lovromazgon) [reply source/iterator/iterator.go:119]: What's strange to me is that the method is not idempotent, if position is `nil` the result will always be different 😕

File: source/iterator/iterator_test.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+	snapshot, tracking, trigger := underTest.HelperObjects()
+	is.Equal(underTest.Position.SnapshotTable, snapshot)
+	is.Equal(underTest.Position.TrackingTable, tracking)
+	is.Equal(underTest.Position.Trigger, trigger)
+}
+
+func checkHelperObject(is *is.I, s string, prefix string) {
+	is.True(strings.HasPrefix(s, prefix))
+	is.True(len(s) < 30)
```
Comment (lovromazgon) [inline source/iterator/iterator_test.go:51]: Nice, I like that we check this 👍

File: source/iterator/iterator.go (hunk 1 above)
Comment (hariso) [reply source/iterator/iterator.go:119]: We may have different understanding and different expectations, but I believe it's not actually expected that this is idempotent. For a given input config (table name etc.), different runs need to create different tracking tables so that they're independent.

Comment (hariso) [reply source/iterator/iterator.go]: Updated!

**Comment by hariso** [comment]
@lovromazgon I was free to make the tracking prefix configurable in this PR.

File: source/iterator/snapshot.go [hunk 4]
```diff
@@ -166,7 +166,7 @@ func (iter *Snapshot) Next(_ context.Context) (sdk.Record, error) {
 
 	transformedRowBytes, err := json.Marshal(transformedRow)
 	if err != nil {
-		return sdk.Record{}, fmt.Errorf("marshal row: %w", err)
+		return sdk.Record{}, fmt.Errorf("ToSDK row: %w", err)
```
[the remark this answers is not in the document]
Comment (hariso) [reply source/iterator/snapshot.go:169]: Good catch, it wasn't intentional. It looks GoLand went a bit too far.

File: source/iterator/position.go [hunk 5]
```diff
@@ -35,7 +35,10 @@ type Position struct {
 	Mode Mode `json:"mode"`
 
 	// LastProcessedVal represents the last processed value from ordering column.
-	LastProcessedVal any `json:"last_processed_val"`
+	LastProcessedVal any    `json:"last_processed_val"`
+	TrackingTable    string `json:"tracking_table"`
+	Trigger          string `json:"trigger"`
+	SnapshotTable    string `json:"snapshot_table"`
```
[the remark this answers is not in the document]
Comment (hariso) [reply source/iterator/position.go:41]: That's true. The reason why I've put it like this is that, at least in my experience, persisting just a part of some information and then reconstructing all of the info from it saves some space but complicates migrations when the pattern changes.

File: source/iterator/iterator.go [hunk 6]
```diff
@@ -104,30 +105,75 @@ type Iterator struct {
...
+	return p
+}
+
+// setHelperObjects sets the names of the helper tables and triggers we use:
+// the snapshot table, the tracking table and the trigger name.
+func (p *Params) setHelperObjects(prefix string) {
+	if p.Position != nil {
+		p.SnapshotTable = p.Position.SnapshotTable
+		p.TrackingTable = p.Position.TrackingTable
+		p.Trigger = p.Position.Trigger
```
[the remark this answers is not in the document]
Comment (hariso) [reply source/iterator/iterator.go:146]: > So if I have an existing pipeline, I change the tracking prefix and then restart the pipeline, it will have no effect? confused

Good point... IMHO, updating the configurations should be a separate task, because there's a few cases we need to cover there. It's likely outside the scope of this PR. E.g. when you change the prefix, there still might be data in the old tracking table, so we need to figure out what to do with it. On the other hand, if we're not reading the tracking table name from the position, we would read it from the config. We can have a case where the config changed from a custom prefix to an auto-generated one. Currently, the connector README points out that manual changes to the tracking tables can cause unexpected behavior, so unless you disagree it looks like it's better to tackle this separately.





> And I have no control over the random postfix added to the table, so if I somehow lose my Conduit state I have no way to force Conduit to use an existing tracking table that is already full of data.

Unless I'm missing something, even in that case it wouldn't be possible to resume from exactly the same place where the pipeline stopped. We need to know where exactly in the snapshot or the tracking table did the connector stop. That is in the position, which in this case would be lost with the rest of the Conduit state.

[the remark this answers is not in the document]
Comment (hariso) [reply source/iterator/iterator.go:146]: > Then changing the prefix would have an effect and the user could be in full control over the helping object names.

For the record, we discussed this topic of configuration at a team sync. We want to make the above-mentioned full control over helper object names possible. We also want to make it possible to specify only the prefix and let the connector handle the uniqueness by adding a random (requirement of our internal platform). That means we'll have 3 options:
1. Names completely generated by the connector, i.e. no config provided by a user (e.g. conduit_snapshot_123)
2. Prefix configured by a user (e.g. `prefix=foobar` would give `snapshot_table=foobar_snapshot_456` and `tracking_table=foobar_tracking_456`).
3. Full control over all names (e.g. `snapshot_table=table_s`, `tracking_table=table_t`)

File: source/iterator/position.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (hariso) [reply source/iterator/position.go:41]: Another reason appeared meanwhile: given the decision from our team sync to make it possible to have custom names, it makes the above decision basically mandatory.

File: source/iterator/iterator.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (lovromazgon) [reply source/iterator/iterator.go:137]: > when you change the prefix, there still might be data in the old tracking table, so we need to figure out what to do with it

This will be easier to implement once we have connector lifecycle events. The connector will know when the config changes and react accordingly. If the connector "owns" the helper objects I'd expect it to rename them (or delete and recreate) and ensure that it doesn't lose existing data in the tracking table. If the objects were created by someone else, I'd expect the connector to simply switch to the new ones, regardless what was there before. Either way, I don't think this is strictly required in this issue, but I would like us to consider the bigger picture to make sure the configuration we are adding is what we want.

That reminds me of another thing - the connector currently always creates the table and trigger, but what if it doesn't have permission to do this? We should probably allow the user to create the objects and just use them if they already exist.

> even in that case it wouldn't be possible to resume from exactly the same place where the pipeline stopped.

Yes, we wouldn't necessarily start at the _exact_ same row, we might create some duplicated records. Looking at the code, the connector deletes processed rows from the tracking table every 5 seconds, so in the worst case we would reprocess 5 seconds worth of rows. I'd expect that in a graceful stop the connector removes all successfully processed rows from the tracking table, so unless the pipeline was stopped forcefully, we should still be able to resume exactly where we left off.¹

From how I understand the connector, Oracle is pushing data to the tracking table via the trigger and the connector is just consuming whatever gets written to the table. Once a record is processed, the corresponding row is eventually removed. This seems like the position stored in Conduit doesn't play much of a role, the connector is tracking its own progress in an Oracle table. So now I'm asking myself why do we even need to read anything from the record position? It seems like the connector should simply read the rows from the tracking table and ensure they are deleted when records are acked, then it will always resume at the correct position.

**TL;DR:** I have a couple of concerns about the current behavior of the connector which are clearly out of scope of this issue. At the same time, I'd like to avoid making breaking changes, so I think we should take a moment to define the end goal regarding configuration options, the behavior we expect from the connector for these options, and how the connector behaves if they change. Then we can be sure that the options and behavior we are introducing here are here to stay.

---

¹ Sidenote: I don't think that's the case right now. The last cleanup is triggered in [`Teardown`](https://github.com/conduitio-labs/conduit-connector-oracle/blob/a42b9bc1add95acde578cce609d06c52e8287a74/source/source.go#L157), which sends a [signal to the cleanup routine](https://github.com/conduitio-labs/conduit-connector-oracle/blob/a42b9bc1add95acde578cce609d06c52e8287a74/source/iterator/cdc.go#L228), but the `ctx` used in the [cleanup routine](https://github.com/conduitio-labs/conduit-connector-oracle/blob/a42b9bc1add95acde578cce609d06c52e8287a74/source/iterator/cdc.go#L324) is coming from `Open`, and that context is already closed at the time `Teardown` is called.

File: config/source.go [hunk 7]
```diff
@@ -143,3 +146,30 @@ func ParseSource(cfgMap map[string]string) (Source, error) {
 
 	return cfg, nil
 }
+
+func (s *Source) setHelperObjects(cfgMap map[string]string) {
+	s.setDefaultHelperObjects()
```
Comment (lovromazgon) [inline config/source.go:151]: Sidenote: If we change the behavior to take these names from the position _only if_ the configuration doesn't specify them, then we'll have to know if the config was actually empty. No issue now, but we might have to change this in the future.

**Review by lovromazgon — APPROVED** [review]
👍

File: config/source.go (hunk 7 above)
Comment (hariso) [reply config/source.go:151]: Yes, you're correct about that. The reason why I put this here is that I wanted the result of `Parse` to be a complete configuration.
